include/zephyr/bluetooth/l2cap.h

```diff
@@ -279,6 +279,18 @@ struct bt_l2cap_chan_ops {
 	 */
 	void (*encrypt_change)(struct bt_l2cap_chan *chan, uint8_t hci_status);
 
+	/** @brief Channel alloc_seg callback
+	 *
+	 *  If this callback is provided the channel will use it to allocate
+	 *  buffers to store segments. This avoids wasting big SDU buffers with
+	 *  potentially much smaller PDUs.
+	 *
+	 *  @param chan The channel requesting a buffer.
+	 *
+	 *  @return Allocated buffer.
```

Comment:
.."Or NULL to fallback on the SDU allocator."? Or is this behavior unintentional? If it's unintentional, we should log an error.

Reply:
behavior is unintentional, I think we should log an error.

Reply:
I moved the check into an assert, lemme know if you don't like it.

The behavior will be that without asserts enabled (which you shouldn't do bc of the stack relying on it), segments will be allocated from the global pdu pool.

Reply:
no, it's in the right place:
@alwa-nordic was saying that I had undocumented behavior where if the allocator cb returns NULL, the segment would tentatively be allocated from the same pool as the SDU.
I decided to make it mandatory to return a valid buffer, added a line to this comment and added an assert to verify that.

I have already changed the line with the assert in the bsim test, that's the one you both had problems with.